network/firewall.py

```diff
@@ -0,0 +1,396 @@
+                prio = 100 if hook == "postrouting" else 0
+                if not cls.check_if_table_exists("ip", table):
+                    commands.append({"add": {"table": {"family": "ip", "name": table}}})
+                new_chain = {
+                    "chain": {
+                        "family": "ip",
+                        "table": table,
+                        "name": chain,
+                        "type": table,
+                        "hook": hook,
+                        "prio": prio,
+                    }
+                }
+                commands.append({"add": new_chain})
+                chains.append(new_chain)
+            elif len(chains) > 1:
+                logger.critical(
+                    f"Unsupported: Multiple base chains already defined for hook {hook}."
+                )
+                # TODO: gracefully exit
+            base_chains[hook] = chains.pop()
+
+        cls.forwarding_info = base_chains["forward"]["chain"]
+        cls.postrouting_info = base_chains["postrouting"]["chain"]
+
+        cls.add_chain("ip", cls.postrouting_info["table"], "aleph-supervisor-nat")
+        commands.append(
+            {
+                "add": {
+                    "rule": {
+                        "family": "ip",
+                        "table": cls.postrouting_info["table"],
+                        "chain": cls.postrouting_info["name"],
+                        "expr": [{"jump": {"target": "aleph-supervisor-nat"}}],
+                    }
+                }
+            }
+        )
+
+        cls.add_chain("ip", cls.forwarding_info["table"], "aleph-supervisor-filter")
+        commands.append(
+            {
+                "add": {
+                    "rule": {
+                        "family": "ip",
+                        "table": cls.forwarding_info["table"],
+                        "chain": cls.forwarding_info["name"],
+                        "expr": [{"jump": {"target": "aleph-supervisor-filter"}}],
+                    }
+                }
+            }
+        )
+        commands.append(
+            {
+                "add": {
+                    "rule": {
+                        "family": "ip",
+                        "table": cls.forwarding_info["table"],
+                        "chain": "aleph-supervisor-filter",
+                        "expr": [
+                            {
+                                "match": {
+                                    "op": "in",
+                                    "left": {"ct": {"key": "state"}},
+                                    "right": ["related", "established"],
+                                }
+                            },
+                            {"accept": None},
+                        ],
+                    }
+                }
+            }
+        )
+
+        cls.execute_json_nft_commands(commands)
+        cls.firewall_has_been_setup = True
```

Comment:
Do I understand that there is a class `Firewall` that only has class methods and whose state is modified ? That class uses a global `nft` variable.
Why not use normal methods and mutate an instance instead ? And create `self.nft` in the `__init__` of the class ?

Reply:
Yes. As noted above, nft should just move into the single function even. And the one other function that still uses it (get_existing_nftables_ruleset() ) should be rewritten to use the other function as well.